### PR TITLE
Ensuring that errors encountered for state transitions are handled with redirect responses for Works

### LINKED ISSUE #527: Describe Form: Limit values of contributorType for depositor only
## Description 
The [(7.a) contributorType](https://support.datacite.org/docs/datacite-metadata-schema-v44-recommended-and-optional-properties#7a-contributortype) property for Datacite schema has numerous values. We want to limit the number of values for the depositor only. The moderator role (e.g., curators and others with admin access) will be able to view and select from the complete list of choices. When a moderator has chosen a value that is restricted in the depositor's list, that value will be locked for that specific contributor only. The depositor would discuss changes for the moderator to make if needed.

### List of values the depositor will see:
- Researcher (default)
- DataCollector
- DataCurator
- ProjectLeader
- ProjectMember

## User Story
Dr. Jonas Salk, a depositor, is adding a contributor: Ruby Hirose. Dr. Salk will see the default is "Researcher" which happens to be the correct selection, so he will stop there. Dr. Salk than adds a new contributor: Tyler Wade. He also keeps the default "Researcher" for Tyler Wade. Dr. Salk is consulting with the assigned curator, Neggin. Neggin notices that Tyler Wade is a different contributor type that is not available on the depositor's list. Neggin will select "DataManager" from the complete list of values that is available to moderators. This action will lock the selection for this contributor (Tyler Wade) only. Dr. Salk may ask Neggin to further change the selection for Tyler Wade, but he cannot change it. Dr.Salk may still change the selection for Ruby Hirose or other contributors he may add to the record.

## Acceptance Criteria

Depositor values are limited to:
- [ ] Researcher (default)
- [ ] DataCollector
- [ ] DataCurator
- [ ] ProjectLeader
- [ ] ProjectMember  

Other:
- [ ] The depositor's default is "Researcher"
- [ ] Moderators have access to the complete list of values
- [ ] When a moderator chooses a value that is not in the depositor's list, the specific contributor will become locked, so that only the moderator may change it.
- [ ] Locking one contributor does not affect other contributors in the list or those that may be added by the depositor.

### LINKED ISSUE #1527: Errors on approval throw exception and do not create a nice error message
The following error is showing up in honey badger and the user gets an "Oops something went wrong error"

to reproduce click the approve button from: https://pdc-describe-staging.princeton.edu/describe/works/374

[pdc_describe/staging] Notice: Invalid Approve: Event 'approve' cannot transition from 'awaiting_approval'. Failed callback(s): [:valid_to_approve]. errors: Uploads must be present for a work to be approved

## Backtrace

    line 358 of [PROJECT_ROOT]/app/controllers/works_controller.rb: rescue in rescue_aasm_error
    line 351 of [PROJECT_ROOT]/app/controllers/works_controller.rb: rescue_aasm_error

[View full backtrace and more info at honeybadger.io](https://app.honeybadger.io/projects/99214/faults/100336746)

Acceptance criteria:

- [ ] any AASM errors get captured when you hit the approve button and a nice error is given to the user
- [ ] In this specific case tell the user they can not approve the work with no files attached

### PR DESCRIPTION
**Pull request by jrgriffiniii**
Resolves #1527

### CODE REVIEW

**Comment by jrgriffiniii** [comment]
There is one remaining failing test, and I am struggling to understand how this failure was introduced, and why this is exactly failing with the form submission:

https://app.circleci.com/pipelines/github/pulibrary/pdc_describe/4468/workflows/2cbd3e36-5697-4202-bf44-250292ebfe30/jobs/16478

image

**Comment by jrgriffiniii** [comment]
This seems to be related to the form submission executed by Capybara/Selenium, specifically at the following line:

https://github.com/pulibrary/pdc_describe/pull/1548/files#diff-c37a48a9d764aaae554d5b61caa1176aa19ba56425fc7a4fa9d2aaaa524f68a1R120
I am troubleshooting this and finding that the selected `<option>` elements do not seem to be properly selected:

```bash
find("tr:last-child select[name='related_objects[][related_identifier_type]'] option[selected]")['innerHTML']
=> ""
```

**Comment by jrgriffiniii** [comment]
Inspecting the `PATCH` parameters for this test yields the following:

```bash
[4] pry(#<WorksController>)> unescaped
=> "_method=patch&work[work_id]=&doi_read_only=10.34770/doc-1&title_main=Sowing the Seeds for More Usable Web Archives: A Usability Study of Archive-It&description=Download the README.txt for a detailed description of this dataset's content.&rights_identifiers[]=GPLv3&rights_identifiers[]=CC BY&creators[][given_name]=Samantha&creators[][family_name]=Abrams&creators[][orcid]=&creators[][affiliation]=&creators[][ror]=&creators[][given_name]=Alexis&creators[][family_name]=Antracoli&creators[][orcid]=&creators[][affiliation]=&creators[][ror]=&creators[][given_name]=Rachel&creators[][family_name]=Appel&creators[][orcid]=&creators[][affiliation]=&creators[][ror]=&creators[][given_name]=Celia&creators[][family_name]=Caust-Ellenbogen&creators[][orcid]=&creators[][affiliation]=&creators[][ror]=&creators[][given_name]=Sarah&creators[][family_name]=Dennison&creators[][orcid]=&creators[][affiliation]=&creators[][ror]=&creators[][given_name]=Sumitra&creators[][family_name]=Duncan&creators[][orcid]=&creators[][affiliation]=&creators[][ror]=&creators[][given_name]=Stefanie&creators[][family_name]=Ramsay&creators[][orcid]=&creators[][affiliation]=&creators[][ror]=&creators[][given_name]=Dionna&creators[][family_name]=Orn&creators[][orcid]=&creators[][affiliation]=&creators[][ror]=&keywords=&funders[][funder_name]=&funders[][award_number]=&funders[][award_uri]=&funders[][ror]=&related_objects[][related_identifier]=https://related.example.com&related_objects[][related_identifier_type]=&related_objects[][relation_type]=&contributors[][orcid]=&contributors[][given_name]=&contributors[][family_name]=&contributors[][role]=&organizational_contributors[][ror]=&organizational_contributors[][value]=&organizational_contributors[][type]=DATA_COLLECTOR&publisher=Princeton University&publication_year=2023&doi=10.34770/doc-1&ark=&resource_type=Dataset&resource_type_general=Dataset&version_number=1&group_id=1&embargo-date=&collection_tags=&commit=Save Work&existing_title_count=1&new_title_count=0&creator_count=8&contributor_count=0&related_object_count=0&work[deleted_files_count]=0&doi_read_only=10.34770/doc-1&wizard=true&user_orcid=&user_given_name=Dionna&user_family_name=Orn"
```

Capybara/Selenium does not seem to be submitting the HTML form values for `related_objects[][related_identifier_type]` and `related_objects[][relation_type]`:

```bash
related_objects[][related_identifier]=https://related.example.com&related_objects[][related_identifier_type]=&related_objects[][relation_type]=&
```

**Comment by jrgriffiniii** [comment]
I apologize for the continued delays on this PR. I have made significant progress with https://github.com/pulibrary/pdc_describe/pull/1548/commits/9c8f6e41664b286e6c4ba4c3dfc2ee9f329be198, as there was a bug in which end-users were able to submit invalid "Additional Metadata" to Works which are within the `draft` state, which could then be persisted (but which should not be).

**Comment by jrgriffiniii** [comment]
I have this stabilized finally, I can look to clean this PR and prepare this for review. Thank you again for your patience.

File: app/models/work.rb [hunk 1]
```diff
@@ -126,6 +126,8 @@ def find_by_ark(ark)
       work.validate_ids
     elsif draft?
       work.valid_to_draft
+      # There are cases where, before the README is attached, invalid metadata may be provided
```
Comment (carolyncole) [inline app/models/work.rb:129]: I think this new set of code will ignore errors in valid_to_draft.  Could these checks be added there instead?

File: spec/system/work_create_spec.rb [hunk 2]
```diff
@@ -78,96 +81,250 @@
...
+      first_creator = work.resource.creators.first
+      expect(first_creator.given_name).to eq("Samantha")
+      expect(first_creator.family_name).to eq("Abrams")
+      last_creator = work.resource.creators.last
+      expect(last_creator.given_name).to eq(user.given_name)
+      expect(last_creator.family_name).to eq(user.family_name)
+    end
+
+    context "when failing to provide the required metadata" do
+      xit "it renders a warning in response to form submissions" do
```
Comment (carolyncole) [inline spec/system/work_create_spec.rb:96]: This should not be pending, as you are removing tests that previously existed

File: spec/system/work_create_spec.rb [hunk 3]
```diff
@@ -78,96 +81,250 @@
...
+        # click_on "Create New"
+        # expect(page).to have_content("Must provide a title")
+        # expect(page).to have_content("Must provide at least one creator")
+      end
+    end
+
+    context "when failing to provide the given name and family name for the creator" do
+      xit "it renders a warning in response to form submissions" do
```
Comment (carolyncole) [inline spec/system/work_create_spec.rb:104]: This should not be pending, as you are removing tests that previously existed

File: spec/system/work_create_spec.rb [hunk 4]
```diff
@@ -78,96 +81,250 @@
...
       click_on "Save Work"
-      click_on "Back"
-      expect(page).to have_content(description)
-      click_on "Save Work"
-      expect(page).to have_content("Please upload the README")
-      expect(page).to have_button("Continue", disabled: true)
-      path = Rails.root.join("spec", "fixtures", "files", "readme.txt")
-      attach_file(path) do
-        page.find("#patch_readme_file").click
+      page.save_screenshot
```
Comment (carolyncole) [inline spec/system/work_create_spec.rb:123]: I'm guessing this is cruft
```suggestion
```

File: spec/system/work_create_spec.rb [hunk 5]
```diff
@@ -78,96 +81,250 @@
...
+        work.reload
+        stub_s3 data: [FactoryBot.build(:s3_readme, work:)]
+
+        readme = Readme.new(work, user)
+        expect(readme.file_name).not_to be nil
+        expect(readme.file_name).to eq("README.txt")
+      end
+
+      context "when a README has already been attached to a Work" do
+        xit "attempting to upload another README renders an error" do
```
Comment (carolyncole) [inline spec/system/work_create_spec.rb:243]: Is there a reason for adding a pending test?  We rarely change the pending tests in the future.

**Review by carolyncole — CHANGES_REQUESTED** [review]
@jrgriffiniii I am surprised that this PR is changing the validation criteria.  That said with your changes to the tests and the changes to the code it is hard to see if the code still works correctly.  I believe that the pending tests might be the ones that prove that work.valid_to_draft is being ignored at line 128.  I'm happy to be wrong, but I would like to see those pending test implemented to prove your change has not negative consequences.

**Review by carolyncole — APPROVED** [review]

**Pull request merged**